acinclude/os-deps.m4

```diff
@@ -968,3 +968,29 @@ AC_DEFUN([SQUID_CHECK_BROKEN_SOLARIS_IPFILTER],[
   ])
 
 ])
+
+AH_TEMPLATE(STD_UNORDERED_MAP_ALLOCATOR_WORKS,
+  [OpenBSD brokenness: for some reason it messes up const-ness in std::unordered_map if a allocator is supplied])
+AC_DEFUN([SQUID_CHECK_UNORDERED_MAP_ALLOCATOR_WORKS],[
+  AC_CACHE_CHECK([if std::unordered_map can accept an allocator],
+    squid_cv_std_unordered_map_allocator_works,
+    AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
+#include <memory>
+#include <unordered_map>
+template <class T>
+class SomeAllocator : public std::allocator<T>
+{};
+
+using Element = std::pair<int, uint64_t>;
+using Map = std::unordered_map<int, uint64_t, std::hash<int>, std::equal_to<int>, SomeAllocator<Element> >;
+
+static Map testMap;
```

Comment:
> PR description: OpenBSD's STL has a const-correctness bug in std::unordered_map when supplying an allocator.

AFAICT, the bug is in Squid code, not OpenBSD or STL. The proposed ./configure test (besides other problems) replicates the same Squid bug and, hence, fails with [many](https://godbolt.org/z/nqso1xqW3) [compilers](https://godbolt.org/z/TbrP4MfjG) outside of OpenBSD environment.

[std::unordered_map](https://en.cppreference.com/w/cpp/container/unordered_map), like other STL containers, requires that the container and the allocator have the same value_type. For std::unordered_map, value_type is `std::pair<const Key, T>`. We are providing an allocator with value_type `std::pair<Key, T>`. Compiler errors[^1] in my tests are actually pretty clear about that requirement -- "unordered container must have the same value_type as its allocator".

Please fix Squid code to use the right allocator value type in all containers. For example, HeaderValueCountsElement should use `std::pair<const SBuf, uint64_t>`. ClpMap::IndexItem should use `std::pair<const Key, EntriesIterator>`.

[^1]: The original error is missing from the PR description. Please add (its abridged version)!

I do not know why some compilers allow the mismatch in some environments, but that is not important.

Reply:
Aha! got it; i had tried fiddling with const-ness in the element but I thought it would not need a separate type

```
using Element = std::pair<int, uint64_t>;
using AllocatorElement = std::pair<const int, uint64_t>;
using Map = std::unordered_map<int, uint64_t, std::hash<int>, std::equal_to<int>, SomeAllocator<AllocatorElement> >;
```
works; changing to that instead. Thanks for doing the analysis!